Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -203,146 +203,180 @@ function xASL_wrp_ResampleASL(x)
+            % we need the PLD vector now for the pairwise subtractions only            
+            initialPLD_PWI4D = initialPLD(1:2:end);
+            % Do the same for LabDur
+            LabDuration_PWI4D = LabDuration(1:2:end);
+        else % the same but then without subtraction
+            PWI4D = ASL_im;
             
-        else
-            PWI = xASL_stat_MeanNan(ASL_im, 4); % singlePLD PWI
-            % Save PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWI, 32, false);
-        end            
+            initialPLD_PWI4D = initialPLD;
+            LabDuration_PWI4D = LabDuration;
+        end
+       
+        %% 2) Create PWI3D
+
+		% After averaging across PLDs, we'll obtain these unique PLDs+LD combinations
+		% indexAverage_PLD_LabDur lists for each original position to where it should be averaged
+		[~, ~, iUnique_PLD_LabDur] = unique(unique([initialPLD_PWI4D, LabDuration_PWI4D], 'stable', 'rows'), 'stable', 'rows');
+
+        % MultiPLD-multiLabDur PWI3D after averaging
+        for iPLD_LabDur = 1:max(iUnique_PLD_LabDur)
+            PWI3D(:, :, :, iPLD_LabDur) = xASL_stat_MeanNan(PWI4D(:, :, :, iUnique_PLD_LabDur == iPLD_LabDur), 4); % Averaged PWI4D 
+        end
+
+        %% 3) Create PWI
+        % We create a dummy CBF image for registration purposes
+        % The earliest echo, and latest PLD-labeling duration combination
+        % are the best for this, having most SNR and CBF-weighting,
+        % respectively
+
+        PWI = PWI3D(:, :, :, iPLD_LabDur==max(iUnique_PLD_LabDur));
+
     end
+    
+    
+    % =====================================================================
+    % D) Save subtracted to disk
+
+    % Save PWI (single volume)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI{iSpace}, PWI, 32, false);
+
+    % Save PWI3D (averaged volume for each PLD-labdur combination)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI3D{iSpace}, PWI3D, 32, false);
+    
+    % Save PWI4D (subtracted only, not yet averaged)
+    fprintf('%s\n', PathPWI4D{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI4D{iSpace}, PWI4D, 32, false);    
 end
 
 
 %% ------------------------------------------------------------------------------------------
 % 9. Save PWI NIfTI & time-series-related maps (SD, SNR)
 MaskIM = xASL_io_Nifti2Im(fullfile(x.D.MapsSPMmodifiedDir, 'rgrey.nii'));
 MaskIM = MaskIM>(0.7*max(MaskIM(:)));
-PWI = xASL_stat_MeanNan(ASL_im, 4); % repeat this here, in case it is skipped above for single volumes (e.g. GE 3D spiral)
 
-if size(ASL_im, 4)>3
+if size(PWI4D, 4)>3
+    %% PM: here we need to use the positions for the earliest echo, and the latest PLD-labdur
+    
     % Save SD & SNR maps
-    SD = xASL_stat_StdNan(ASL_im, 0, 4);
-    SNR = PWI./SD;

```

Comment:
This needs to be improved yet. I will skip the SNR (we don't really use it I believe) and only calculate SD when a TE_PLD_LD combination has more than 2 volumes.

Reply:
You don't skip it, you save it. And we had it previously as a 3D file, now we will have a 4D file. Can you change this back to PWI3D?

Reply:
We use SNR, we save it.
OK - waiting for the improvement

Reply:
A) We want SD for each TE/PLD/LD combination that has a significant number of volumes. E.g., if you have two PLDs with each 20 volumes, you want the SD for both. If you have 8 TEs with each 3 volumes, this might be an overkill. So we should think what we deem helpful.
B) We want to add other stats here, e.g., mean/SD/outlier mapping, maximal and minimal MIP might also give us insight over all TEs, fitting all TEs without any assumptions might also give us insight if things are wrong in the data or in the interaction of the data with a model. 
C) We never used SNR, we did use SD. So, I propose to remove SNR for space/computation efficiency. But that's not the most important point here

We might also skip this section here, and do this all in the QC wrapper.

Shall I for now, keep the original code here (so revert to PWI, which is highest PLD/LD, shortest TE etc)?
And move our wishlist to the separate QC issue. Makes it also easier for #1543

Reply:
A) I don't think we need SD for each TE - that will eat extremely lot of space, especially if you save that to standard space. 
B) Sure, but not for the entire PWI4D. We can add more stats later when we decided which ones we need, but now it's an overkill.
C) I'm fine removing SNR. But then you have to remove it completely.

we can for sure do that in QC. 
Reverting to PWI and adding to a wish list is fine.

Reply:
We can see later which QC parameters are useful and which aren't, but we already delete the standard space PWI4D at the end of the xASL_module_ASL. We keep native space PWI4D (we need it for merging). Do you agree?

Reply:
This appears to be fixed and a single SD is calculated, right? So we can resolve.